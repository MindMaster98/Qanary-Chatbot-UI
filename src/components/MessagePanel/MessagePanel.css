#messagePanel {
    position: relative;
    height: 100%;
    width: 100%;
    overflow-y: auto;
}

#messagePanel *:first-child {
    margin-top: 2vh;
}

/*
#messagePanel::before {
    content: 'Covid-19 Chatbot';
    text-align: center;
    display: inline-block;
    font-size: large;
    height: 3vh;
    width: 100%;
    margin-top: 3vh;
    margin-bottom: 3vh;
    color: var(--dark-teal);
}
*/
