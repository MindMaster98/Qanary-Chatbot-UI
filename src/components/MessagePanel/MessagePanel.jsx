import React, { Component } from "react";
import { Message, RichMessage } from "..";
import "./MessagePanel.css";


/**
 * @description Contains and displays all messages as the correct components
 * 
 * @property {array} messages all message objects, as specified by the typedef in `Message.jsx`, possibly with a `visualization` property containing data on, for instance, buttons
 */
export default class MessagePanel extends Component {
  componentDidUpdate() {
    // this component updates when new messages are added
    // so with every new message, a scroll to the newest message is triggered
    window.requestAnimationFrame(() => {
      const messagePanel = document.getElementById("messagePanel");
      messagePanel.scrollTop += messagePanel.scrollHeight;
    });
  }

  render() {
    const messageElements = this.props.messages?.map?.((message, i) =>
    // i is the shortest unique identifier in this case and the content will not be shuffled, so i is constant
    !!message.visualization ? (
      <RichMessage key={i} id={i} messageObject={message} />
    ) : (
      <Message key={i} messageObject={message} />
    )
  )
    return (
      <div tabIndex="-1" id="messagePanel">
        {!!messageElements?.length > 0 ? messageElements : null}
      </div>
    );
  }
}
