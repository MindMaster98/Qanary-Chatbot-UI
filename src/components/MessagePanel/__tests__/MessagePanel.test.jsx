import renderer from "react-test-renderer";
import MessagePanel from "../MessagePanel";

// Snapshot tests

it("renders correctly with no props", () => {
  const tree = renderer.create(<MessagePanel />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders correctly with one valid message", () => {
  const tree = renderer.create(<MessagePanel messages={[
    {
      isReply: false,
      loadedSuccessfully: true,
      text: "test",
      followUpNeeded: false,
      time: "13:37",
    }
  ]} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders correctly with multiple distinct valid messages", () => {
  const tree = renderer.create(<MessagePanel messages={[
    {
      isReply: false,
      loadedSuccessfully: true,
      text: "test",
      followUpNeeded: false,
      time: "13:37",
    },
    {
      isReply: false,
      loadedSuccessfully: true,
      text: "test2",
      followUpNeeded: false,
      time: "13:37",
    },
    {
      isReply: false,
      loadedSuccessfully: true,
      text: "test3",
      followUpNeeded: false,
      time: "13:37",
    }
  ]} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders correctly with multiple equal valid messages", () => {
  const tree = renderer.create(<MessagePanel messages={[
    {
      isReply: false,
      loadedSuccessfully: true,
      text: "test",
      followUpNeeded: false,
      time: "13:37",
    },
    {
      isReply: false,
      loadedSuccessfully: true,
      text: "test",
      followUpNeeded: false,
      time: "13:37",
    },
    {
      isReply: false,
      loadedSuccessfully: true,
      text: "test",
      followUpNeeded: false,
      time: "13:37",
    }
  ]} />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders correctly with just empty object", () => {
  const tree = renderer.create(<MessagePanel messages={[{}]} />).toJSON();
  expect(tree).toMatchSnapshot();
});