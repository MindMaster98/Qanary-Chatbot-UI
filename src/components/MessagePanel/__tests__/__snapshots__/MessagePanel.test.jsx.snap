// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly with just empty object 1`] = `
<div
  id="messagePanel"
  tabIndex="-1"
>
  <div
    className="messageContainer"
    tabIndex="-1"
  >
    <div
      className="clickableIcon "
      style={
        Object {
          "cursor": "initial",
          "height": "100px",
          "width": "100px",
        }
      }
      tabIndex="-1"
    >
      <img
        alt="Nutzer"
        src="account_icon_black.svg"
        tabIndex="-1"
      />
    </div>
    <div
      className="messageDataContainer"
      tabIndex="0"
    >
      <div
        className="messageText black-text"
        dangerouslySetInnerHTML={
          Object {
            "__html": "",
          }
        }
      />
      <div
        className="messageTimeSent"
      />
    </div>
  </div>
</div>
`;

exports[`renders correctly with multiple distinct valid messages 1`] = `
<div
  id="messagePanel"
  tabIndex="-1"
>
  <div
    className="messageContainer"
    tabIndex="-1"
  >
    <div
      className="clickableIcon "
      style={
        Object {
          "cursor": "initial",
          "height": "100px",
          "width": "100px",
        }
      }
      tabIndex="-1"
    >
      <img
        alt="Nutzer"
        src="account_icon_black.svg"
        tabIndex="-1"
      />
    </div>
    <div
      className="messageDataContainer"
      tabIndex="0"
    >
      <div
        className="messageText black-text"
        dangerouslySetInnerHTML={
          Object {
            "__html": "test",
          }
        }
      />
      <div
        className="messageTimeSent"
      >
        13:37
      </div>
    </div>
  </div>
  <div
    className="messageContainer"
    tabIndex="-1"
  >
    <div
      className="clickableIcon "
      style={
        Object {
          "cursor": "initial",
          "height": "100px",
          "width": "100px",
        }
      }
      tabIndex="-1"
    >
      <img
        alt="Nutzer"
        src="account_icon_black.svg"
        tabIndex="-1"
      />
    </div>
    <div
      className="messageDataContainer"
      tabIndex="0"
    >
      <div
        className="messageText black-text"
        dangerouslySetInnerHTML={
          Object {
            "__html": "test2",
          }
        }
      />
      <div
        className="messageTimeSent"
      >
        13:37
      </div>
    </div>
  </div>
  <div
    className="messageContainer"
    tabIndex="-1"
  >
    <div
      className="clickableIcon "
      style={
        Object {
          "cursor": "initial",
          "height": "100px",
          "width": "100px",
        }
      }
      tabIndex="-1"
    >
      <img
        alt="Nutzer"
        src="account_icon_black.svg"
        tabIndex="-1"
      />
    </div>
    <div
      className="messageDataContainer"
      tabIndex="0"
    >
      <div
        className="messageText black-text"
        dangerouslySetInnerHTML={
          Object {
            "__html": "test3",
          }
        }
      />
      <div
        className="messageTimeSent"
      >
        13:37
      </div>
    </div>
  </div>
</div>
`;

exports[`renders correctly with multiple equal valid messages 1`] = `
<div
  id="messagePanel"
  tabIndex="-1"
>
  <div
    className="messageContainer"
    tabIndex="-1"
  >
    <div
      className="clickableIcon "
      style={
        Object {
          "cursor": "initial",
          "height": "100px",
          "width": "100px",
        }
      }
      tabIndex="-1"
    >
      <img
        alt="Nutzer"
        src="account_icon_black.svg"
        tabIndex="-1"
      />
    </div>
    <div
      className="messageDataContainer"
      tabIndex="0"
    >
      <div
        className="messageText black-text"
        dangerouslySetInnerHTML={
          Object {
            "__html": "test",
          }
        }
      />
      <div
        className="messageTimeSent"
      >
        13:37
      </div>
    </div>
  </div>
  <div
    className="messageContainer"
    tabIndex="-1"
  >
    <div
      className="clickableIcon "
      style={
        Object {
          "cursor": "initial",
          "height": "100px",
          "width": "100px",
        }
      }
      tabIndex="-1"
    >
      <img
        alt="Nutzer"
        src="account_icon_black.svg"
        tabIndex="-1"
      />
    </div>
    <div
      className="messageDataContainer"
      tabIndex="0"
    >
      <div
        className="messageText black-text"
        dangerouslySetInnerHTML={
          Object {
            "__html": "test",
          }
        }
      />
      <div
        className="messageTimeSent"
      >
        13:37
      </div>
    </div>
  </div>
  <div
    className="messageContainer"
    tabIndex="-1"
  >
    <div
      className="clickableIcon "
      style={
        Object {
          "cursor": "initial",
          "height": "100px",
          "width": "100px",
        }
      }
      tabIndex="-1"
    >
      <img
        alt="Nutzer"
        src="account_icon_black.svg"
        tabIndex="-1"
      />
    </div>
    <div
      className="messageDataContainer"
      tabIndex="0"
    >
      <div
        className="messageText black-text"
        dangerouslySetInnerHTML={
          Object {
            "__html": "test",
          }
        }
      />
      <div
        className="messageTimeSent"
      >
        13:37
      </div>
    </div>
  </div>
</div>
`;

exports[`renders correctly with no props 1`] = `
<div
  id="messagePanel"
  tabIndex="-1"
/>
`;

exports[`renders correctly with one valid message 1`] = `
<div
  id="messagePanel"
  tabIndex="-1"
>
  <div
    className="messageContainer"
    tabIndex="-1"
  >
    <div
      className="clickableIcon "
      style={
        Object {
          "cursor": "initial",
          "height": "100px",
          "width": "100px",
        }
      }
      tabIndex="-1"
    >
      <img
        alt="Nutzer"
        src="account_icon_black.svg"
        tabIndex="-1"
      />
    </div>
    <div
      className="messageDataContainer"
      tabIndex="0"
    >
      <div
        className="messageText black-text"
        dangerouslySetInnerHTML={
          Object {
            "__html": "test",
          }
        }
      />
      <div
        className="messageTimeSent"
      >
        13:37
      </div>
    </div>
  </div>
</div>
`;
